extractor/src/main/scala/org/jetbrains/sbt/extractors/DependenciesExtractor.scala

```diff
@@ -52,38 +72,77 @@ class DependenciesExtractor(projectRef: ProjectRef,
-
-  private def toSbtStructureConfigurations(confs: Seq[sbt.Configuration]): Seq[jb.Configuration] =
-    mapConfigurations(confs.map(c => jb.Configuration(c.name)))
+  private def forAllConfigurations[T](fn: SbtConfiguration => Seq[T]): Seq[(T, Seq[Configuration])] = {
+    dependencyConfigurations
+      .flatMap(conf => fn(conf).map(it => (it, conf)))
+      .groupBy(_._1)
+      .mapValues(_.unzip._2.map(c => Configuration(c.name)))
+      .toSeq
+  }
 
   // We have to perform this configurations mapping because we're using externalDependencyClasspath
   // rather than libraryDependencies (to acquire transitive dependencies),  so we detect
   // module presence (in external classpath) instead of explicitly declared configurations.
-  private def mapConfigurations(configurations: Seq[jb.Configuration]): Seq[jb.Configuration] = {
-    val jbTestConfigurations = testConfigurations.map(c => jb.Configuration(c.name))
-    val cs = configurations.map(c => if (jbTestConfigurations.contains(c)) jb.Configuration.Test else c).toSet
-
-    if (cs == Set(jb.Configuration.Compile, jb.Configuration.Test, jb.Configuration.Runtime)) {
-      Seq(jb.Configuration.Compile)
-    } else if (cs == Set(jb.Configuration.Compile, jb.Configuration.Test)) {
-      Seq(jb.Configuration.Provided)
+  private def mapConfigurations(configurations: Seq[Configuration]): Seq[Configuration] = {
+    val cs = mergeAllTestConfigurations(configurations)
+    if (cs == Set(Configuration.Compile, Configuration.Test, Configuration.Runtime)) {
+      Seq(Configuration.Compile)
+    } else if (cs == Set(Configuration.Compile, Configuration.Test)) {
+      Seq(Configuration.Provided)
     } else {
       cs.toSeq
     }
   }
+
+  /**
+   * This mapping is required in case of situation
+   * where some project depends on another project in two custom configurations: the test one and the one containing sources.
+   * For example in this project {{{
+   * val CustomTest = config("customtest").extend(Test)
+   * val CustomCompile = config("customcompile").extend(Compile)
+   *
+   * val dummy = (project in file("dummy"))
+   * val utils = (project in file("utils"))
+   *  .configs(CustomTest, CustomCompile)
+   *  .settings(
+   *    inConfig(CustomCompile)(Defaults.configSettings),
+   *    inConfig(CustomTest)(Defaults.configSettings))
+   *  .dependsOn(dummy % "customtest;customcompile")
+   * }}}
+   * Without performing the action of this method, `customtest` configuration will be mapped to `Configuration.Test` in
+   * [[org.jetbrains.sbt.extractors.DependenciesExtractor.mapConfigurations]] and then in the scala plugin, scope for this dependency
+   * would be calculated to `TEST` (in [[org.jetbrains.sbt.project.SbtProjectResolver.scopeFor]]) which would not be truth,
+   * because now the scope for `dummy` dependency should be `COMPILE`.
+   *
+   * Check behaviour of this logic when SCL-18284 will be fixed
```

Comment:
tested in `testCustomConfigurationsWithNestedProjectDependencies` in the scala plugin